src/main/java/io/confluent/copycat/hdfs/HdfsWriter.java

```diff
@@ -1,167 +1,292 @@
+              HdfsSinkConnectorConfig.RECORD_WRITER_PROVIDER_CLASS_CONFIG))).newInstance();
+
+      this.context = context;
+      this.assignment = context.assignment();
+      this.avroData = avroData;
+
+      states = new HashMap<>();
+      for (TopicPartition tp: assignment) {
+        states.put(tp, State.RECOVERY_STARTED);
+      }
+      buffer = new HashMap<>();
+      for (TopicPartition tp: assignment) {
+        buffer.put(tp, new LinkedList<SinkRecord>());
       }
-      offsets = readOffsets();
-      writerProviderClass = (Class<RecordWriterProvider>)
-          Class.forName(connectorConfig.getString(HdfsSinkConnectorConfig.RECORD_WRITER_PROVIDER_CLASS_CONFIG));
+      recovered = new HashSet<>();
+
     } catch (IOException e) {
-      throw new CopycatRuntimeException(e);
-    } catch (ClassNotFoundException e) {
-      throw new CopycatRuntimeException("RecordWriterProvider class not found", e);
+      throw new CopycatException(e);
+    } catch (ClassNotFoundException | IllegalAccessException | InstantiationException e) {
+      throw new CopycatException("RecordWriterProvider class not found", e);
     }
   }
 
-
-  public Map<TopicPartition, Long> getPreviousOffsets() {
+  /**
+   * @return get committed offsets of previous task runs.
+   */
+  public Map<TopicPartition, Long> getCommittedOffsets() {
     return offsets;
   }
 
-  public void writeRecord(TopicPartition topicPart, SinkRecord record)
-      throws CopycatException{
+  public void write(Collection<SinkRecord> records) throws IOException {
+    String topic;
+    int partition;
+    for (SinkRecord record: records) {
+      topic = record.topic();
+      partition = record.kafkaPartition();
+      TopicPartition topicPart = new TopicPartition(topic, partition);
+      buffer.get(topicPart).add(record);
+    }
+    for (TopicPartition topicPart: assignment) {
+      execute(topicPart);
+    }
+  }
+
+  public void recover(TopicPartition topicPart) {
     try {
-      RecordWriter<Long, SinkRecord> writer = getWriter(topicPart, record);
-      writer.write(System.currentTimeMillis(), record);
-      if (!offsets.containsKey(topicPart)) {
-        offsets.put(topicPart, record.getOffset() - 1);
-      }
-      updateRecordCounter(topicPart);
-      if (shouldRotate(topicPart)) {
-        rotate(topicPart);
+      switch (states.get(topicPart)) {
+        case RECOVERY_STARTED:
+          pause(topicPart);
+          nextState(topicPart);
+        case RECOVERY_PARTITION_PAUSED:
+          createWAL(topicPart);
+          nextState(topicPart);
+        case WAL_CREATED:
+          applyWAL(topicPart);
+          nextState(topicPart);
+        case WAL_APPLIED:
+          truncateWAL(topicPart);
+          nextState(topicPart);
+        case WAL_TRUNCATED:
+          resetOffsets(topicPart);
+          nextState(topicPart);
+        case OFFSET_RESET:
+          resume(topicPart);
+          nextState(topicPart);
```

Comment:
Should we have a default case as a sanity check if we are only supposed to get into this method when we are in a certain set of states?